src/main/java/VASSAL/configure/ConfigureTree.java

```diff
@@ -284,6 +321,205 @@ protected JPopupMenu buildPopupMenu(final Configurable target) {
+    }
+    return searchNodes;
+  }
+  
+  
+  // Checks a single string against our search parameters
+  public final boolean checkString(String target, String searchString) {
+    if (matchCase) {
+      return target.contains(searchString);
+    } 
+    else {
+      return target.toLowerCase().contains(searchString.toLowerCase());
+    }         
+  }
+  
+  
+  // Checks a node of the tree against our search parameters
+  public final boolean checkNode(DefaultMutableTreeNode node, String searchString) {
+    Configurable c = null;
+    c = (Configurable) node.getUserObject();
+
+    if (matchNames) {
+      String objectName = c.getConfigureName();
+      if (objectName != null && checkString(objectName, searchString)) {
+        return true;
+      }
+    }
+    
+    if (matchTypes) {
+      String className = getConfigureName(c.getClass());
+      if (className != null && checkString(className, searchString)) {
+        return true;
+      }
+    }
+    
+    return false;
+  }
+  
+  
+  // Search through the tree, starting at the currently selected location (and wrapping around if needed)
+  // Compare nodes until we find our search string (or have searched everything we can search)
+  public final DefaultMutableTreeNode findNode(String searchString) {
+    List<DefaultMutableTreeNode> searchNodes = getSearchNodes((DefaultMutableTreeNode)getModel().getRoot());
+    DefaultMutableTreeNode currentNode = (DefaultMutableTreeNode)getLastSelectedPathComponent();
+
+    DefaultMutableTreeNode foundNode = null;
+    int bookmark = -1;    
+
+    if (currentNode != null) {
+      for (int index = 0; index < searchNodes.size(); index++) {
+        if (searchNodes.get(index) == currentNode) {
+          bookmark = index;
+          break;
+        }
+      }
+    }
+
+    for (int index = bookmark + 1; index < searchNodes.size(); index++) {  
+      if (checkNode(searchNodes.get(index), searchString)) {
+        foundNode = searchNodes.get(index);    
+        break;
+      }
+    }
+
+    if (foundNode == null) {
+      for (int index = 0; index <= bookmark; index++) {    
+        if (checkNode(searchNodes.get(index), searchString)) {
+          foundNode = searchNodes.get(index);    
+          break;
+        }
+      }
+    }
+    
+    return foundNode;
+  }  
+  
+  
+  // Finds out how many total nodes match the search string
+  public final int getNumMatches(String searchString) {    
+    List<DefaultMutableTreeNode> searchNodes = getSearchNodes((DefaultMutableTreeNode)getModel().getRoot());
+    int hits = 0;
+
+    for (int index = 0; index < searchNodes.size(); index++) {
+      if (checkNode(searchNodes.get(index), searchString)) {
+        hits++;
+      }
+    }
+    
+    return hits;    
+  }
+  
+    
+  // Search action - runs search dialog box, then searches
+  protected Action buildSearchAction(final Configurable target) {
+    final Action a = new AbstractAction(searchCmd) {
```

Comment:
I would consider extracting this anonymous AbstractAction into it's own class since it's quite long and the ConfigureTree is already a few hundred lines long. Eclipse has decent refactoring support, extracting this as an inner class or even moving into it's own file should be quite easy.

Reply:
Probably above my skill level with high level Java architecture. I tried to get the Eclipse refactor menu to "do anything" but I could only convince it to move a bunch of minor variables into some other class, couldn't figure out how to get it to extract even a single method let alone a group of them. And discovered when poking around with moving the search stuff to another class "by hand" that they call back into the main class for getting names and stuff. I could probably spend an hour or two and make something that "would run" but it would also almost certainly be worse that what's there now. 

Since the real problem isn't the search itself, but the bulkiness of the overall class, I'd suggest it as a refactoring project for a java pro :-)

Reply:
Ok I'll look into making a PR into this PR.

Reply:
Have a look at https://github.com/Cattlesquat/vassal/pull/1